tests/trx_generator/http_client_async.hpp

```diff
@@ -0,0 +1,166 @@
+      req_.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+      req_.set(http::field::content_type, content_type);
+      req_.body() = request_body;
+      req_.prepare_payload();
+
+      // Look up the domain name
+      resolver_.async_resolve(
+          host, port, [self = this->shared_from_this()](beast::error_code ec, auto res) { self->on_resolve(ec, res); });
+   }
+
+   void on_resolve(beast::error_code ec, tcp::resolver::results_type results) {
+      if (ec) {
+         response_callback_(ec, {});
+         return fail(ec, "resolve");
+      }
+
+      // Set a timeout on the operation
+      stream_.expires_after(std::chrono::seconds(30));
+
+      // Make the connection on the IP address we get from a lookup
+      stream_.async_connect(results, [self = this->shared_from_this()](beast::error_code ec, auto endpt) {
+         self->on_connect(ec, endpt);
+      });
+   }
+
+   void on_connect(beast::error_code ec, tcp::resolver::results_type::endpoint_type) {
+      if (ec) {
+         response_callback_(ec, {});
+         return fail(ec, "connect");
+      }
+
+      // Set a timeout on the operation
+      stream_.expires_after(std::chrono::seconds(30));
+
+      // Send the HTTP request to the remote host
+      http::async_write(stream_, req_, [self = this->shared_from_this()](beast::error_code ec, auto bytes_transferred) {
+         self->on_write(ec, bytes_transferred);
+      });
+   }
+
+   void on_write(beast::error_code ec, std::size_t bytes_transferred) {
+      boost::ignore_unused(bytes_transferred);
+
+      if (ec) {
+         response_callback_(ec, {});
+         return fail(ec, "write");
+      }
+
+      // Receive the HTTP response
+      http::async_read(stream_, buffer_, res_,
+                       [self = this->shared_from_this()](beast::error_code ec, auto bytes_transferred) {
+                          self->on_read(ec, bytes_transferred);
+                       });
+   }
+
+   void on_read(beast::error_code ec, std::size_t bytes_transferred) {
+      boost::ignore_unused(bytes_transferred);
+
+      if (ec) {
+         response_callback_(ec, {});
+         return fail(ec, "read");
+      }
+
+      // Write the message to promise result
+      response_callback_(ec, res_);
+
+      // Gracefully close the socket
+      stream_.socket().shutdown(tcp::socket::shutdown_both, ec);
+
+      // not_connected happens sometimes so don't bother reporting it.
+      if (ec && ec != beast::errc::not_connected)
+         return fail(ec, "shutdown");
+
+      // If we get here then the connection is closed gracefully
+   }
+};
+} // namespace http_client_async_details
+
+struct http_request_params {
+   net::io_context& ioc;
+   char const*      host;
+   char const*      port;
+   char const*      target;
+   int              version;
+   char const*      content_type;
+};
+
+inline void initiate_async_http_request(http_request_params& req_params, char const* request_body,
```

Comment:
I think I would call this `async_http_request`

Reply:
Addressed: [f02287e](https://github.com/AntelopeIO/leap/pull/1052/commits/f02287e44535e51b7ae21d1c90db73c40f51d644)